### ISSUE TITLE
Linux Keybase hangs when it can't steal focus

### ISSUE THREAD

**Issue by jbash** [keybase/keybase-issues#3438]
I'm running the Linux Keybase GUI under FVWM and the XFCE panel on Fedora 29.

Any time there's an incoming chat message and I don't have the cursor over the Keybase window, the program hangs completely for probably a minute or two. It ignores all user events and won't repaint its window. Eventually it recovers completely... until the next message comes in. This happens regardless of whether theKeybase  window is in front, obscured, or not open at all.

I suspect that the real triggering condition is not "mouse is not over the window", but "window does not have the keyboard focus and the program can't grab it". I have FVWM configured for focus-follows-mouse, because click-to-focus is morally wrong. Therefore Keybase usually won't have the focus if the mouse isn't over the window. Even more unusually, I've tried to configure FVWM to prevent all programs from stealing focus. Assuming that's working right, which I haven't really tested, Keybase will *not* get the keyboard if it tries to take it.

I get the feeling that it's trying, and reacting poorly to failure.

I'd like to convince you that chat apps shouldn't try to grab the keyboard in the first place. It amounts to a "randomly send a password to a friend every once in a while" feature. However, since every chat app seems to do it, I'm afraid I'll fail to convince you.

... but I'd at least like it if the program didn't actually hang if it can't steal the focus. And maybe you could at least have the *option* of not trying?

My FVWM config file is attached: [config.txt](https://github.com/keybase/keybase-issues/files/3311800/config.txt)

**Comment by heronhaye** [comment]
That's definitely odd behavior. What is your `keybase version`? Could you trigger this behavior and then `keybase log send`?

**Comment by jbash** [comment]
> What is your keybase version?

Client:  4.1.0-20190612201656+952fee6c59
Service: 4.1.0-20190612201656+952fee6c59

> Could you trigger this behavior and then keybase log send?

I would prefer to see what it's going to send before I send anybody any debug logs. The help doesn't say how I can do that (or even *that* I can do it...).

> (By the way, that file shows up as empty for me)

That's odd, because I just downloaded it back from GitHub and it looks fine to me.

**Comment by heronhaye** [comment]
@jbash sure, it will send the contents of `keybase status` and the contents of each of the log files listed there (e.g., `keybase.service.log`). On systems with systemd, it will also send a tail of the systemd user units for `{keybase,kbfs,keybase.gui,keybase-redirector}.service`, and the list of processes running whose name matches `keybase|kbfs`.  If you prefer, you can send redacted versions directly to me over KBFS or email.

>That's odd, because I just downloaded it back from GitHub and it looks fine to me.
Works for me now.

**Comment by jbash** [comment]
I went and found what appear to be the logs and made some extracts. There was a hang on an incoming message that the UI logged at 15:42, and a couple of other things around there. I've included eveything from 15:40 to 15:42.

There was nothing at the relevant times in the "ek" or "kbfs" logs, even after compensating for time zones.

[Keybase.app.txt](https://github.com/keybase/keybase-issues/files/3312055/Keybase.app.txt)
[keybase.service.txt](https://github.com/keybase/keybase-issues/files/3312057/keybase.service.txt)

**Comment by heronhaye** [comment]
@jbash, I don't believe Fedora uses the systemd units actually. You can check with `systemctl --user is-active keybase`. (My understanding is that systemd user manager isn't supported yet, but could be wrong)

**Comment by heronhaye** [comment]
One more question: is this new behavior with this release or have you always had this even in older versions of Keybase?

**Comment by jbash** [comment]
The only unit that has any systemd journal entries other than starts, stops, and reboots is the keybase.gju one. The journal is attached.

The "failed to connect to proxy" messages might correspond to the hangs, but I'm not sure. I don't remember configuring any proxies and don't see anything in the GUI that would let me configure them.

 I think the texture messages are associated with a different bug were stuff in the UI turns black and I have to restart the program.

[systemd-keybase.gui.txt](https://github.com/keybase/keybase-issues/files/3312112/systemd-keybase.gui.txt)

**Comment by jbash** [comment]
I started trying to use the Linux Keybase GUI probably a month or two ago, and it's always done this.

I can try for a different WM or a blank FVWM config, but I'm not confident that either will give meaningful results. A blank FVWM config may not even work. I can also try to *avoid* loading the XFCE panel, which probably launches a bunch of "desktop components".

Speaking of avoiding, My other weirdness is that I usually try *not* to load any "desktop components" that I can avoid, and may be leaving out something that Keybase expects. I've been avoiding that stuff since Gnome and KDE showed up, and I've to be dragged kicking and screaming into each new abomination since about 1989. I mean, dbus is running, but that doesn't mean I *approve* of dbus.

It may take me some time to get closer to vanilla, and even longer to figure out exactly what non-vanilla part triggers this.

**Comment by jbash** [comment]
Oh, and Fedora does use per-user systemd. I think it has for quite a while. In fact you just made me discover a bunch of new user units to try to turn off. :-)

**Comment by heronhaye** [comment]
Hmm, I wonder if it's a problem with notifications. Does `notify-send test` work?

Does the GUI also freeze if you do something like `ls "$(keybase config get -b mountdir)/private/notme"` (which should show a notification for "Permission denied").

**Comment by jbash** [comment]
Hmm. I'm sorry; it looks like I may have been misled and misled you.

It looks like it's not that it hangs when it doesn't have focus at all. It's that it hangs when it receives a message from a user other than me (which of course usually happens when it doesn't have focus). If *I* send the message, even from another device, nothing hangs, regardless of the focus. Should I change the headline?

During each hang, there are about three of the "Failed to connect to proxy" messages in the systemd journal. Each of those messages is associated with an strace log entry like this:

16047 18:00:56.617442 openat(AT_FDCWD, "/home/jbash/.local/share/keybase/fileoutbox/*a-bunch-of-hex*", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)

It looks like it polls for that file several times at five second intervals, then gives up and issues the "Failed to connect" message, then tries several more series of polls. There are a lot of PIDs involved and I'm not sure which threads belong to which processes.

During all of this, it looks like the "gpu-process" is mostly issuing recvmsg calls on a UNIX socket that's connected to I know not what, and failing with EAGAIN.

**Comment by jbash** [comment]
"notify-send test" does nothing for about 30 seconds, produces no output, and exits with status 0.

The GUI *does* freeze on the "ls" command. The "ls" command itself comes back fairly quickly with permission denied.

**Comment by heronhaye** [comment]
Ok, that sounds like an issue with notifications then. Do you have libnotify and a notification server like dunst?

**Comment by jbash** [comment]
I do have libnotify. I did not personally install it, nor did I know that it existed. Yeat another "desktop component". As far as I know I don't have a notification server running. I probably don't *want* one, but I'll try installing one and let you know the result.

**Comment by jbash** [comment]
With dunst installed and running, `notify-send test` comes back immediately, and neither `ls "$(keybase config get -b mountdir)/private/notme"` hangs the Keybase GUI. None of the above creates a notification (the chat window was already on the screen when the incoming message arrived).

I updated the headline; I assume that's what you guys would want.

This ceases to be urgent for me since I have a workaround. I won't close it because I assume you don't want to actually hang like that no matter what.

**Comment by jbash** [comment]
Oh, and thank you for being so responsive and putting up with my bad guesses.
